cmd/forwarder/auth.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+func newAuth(config AuthConfig, registry metrics.Registry) (*BasicAuth, error) {
+	var err error
+
+	if config.RedisUrl != "" && config.RedisKey == "" {
+		err = errors.New("RedisKey must be set if RedisUrl is set")
+		return nil, err
+	}
+
+	if config.RedisUrl == "" && config.Tokens == "" {
+		err = errors.New("At least one of RedisUrl or Tokens must be set.")
+		return nil, err
+	}
+
+	result, err := NewBasicAuthFromString(config.Tokens, config.HmacKey, registry)
+
+	if err != nil {
+		return result, err
+	}
+
+	pChanges := metrics.GetOrRegisterCounter("log-iss.auth_refresh.changes", registry)
+	pFailures := metrics.GetOrRegisterCounter("log-iss.auth_refresh.failures", registry)
+	pSuccesses := metrics.GetOrRegisterCounter("log-iss.auth_refresh.successes", registry)
+
+	if config.RedisUrl != "" {
+		// Parse redis db out of url
+		u, err := url.Parse(config.RedisUrl)
```

Comment:
Instead of all of this see: https://godoc.org/github.com/go-redis/redis#ParseURL